syntax = "proto3";
package persona;

message Persona {
	repeated DNA DNAs = 1;
	string physical = 2;
	Age age = 3;
	
	message DNA {
		string DNA = 1;
		repeated Input inputs = 2;
		repeated Layer layers = 3;
		repeated Output outputs = 4;
		repeated Connection connections = 5;
		
		message Input {
			string inputLayerName = 1;
			Transform inputTransform = 3;
			repeated Connection connections = 5;
		}

		message Layer {
			string layerName = 1;
			repeated Size layerSize = 2;
			repeated Connection connections = 3;
			
			oneof SubLayer {
				LayerConvolution layerConvolution = 4;
				LayerActivation layerActivation = 5;
				LayerDropout layerDropout = 6;
			}		
		}
		
		message LayerConvolution {
			uint64 convolutionDimension = 1;
			uint64 filters = 2;
			repeated uint64 kernelSize = 3;
			string borderMode = 4;
			repeated uint64 inputShape = 5;
		}
		
		message LayerActivation {
			string activationType = 1;
		}

		message LayerDropout {
			double dropPercentage = 1;
		}
		
		message Output {
			string outputLayerName = 1;
			string outputLayerType = 2;
			Transform outputTransform = 3;
			repeated Connection connections = 4;
		}
		
		message Transform {
			string transformerName = 1;
			string informationType = 2;
			repeated Size transformSize = 3;
			repeated Parameter transformParam = 4;
		}
		
		message Connection {
			string sourceLayerName = 1;
			string destinationLayerName = 2;
		}
		
		message Size {
			uint64 dimension = 1;
			uint64 dimensionSize = 2;
		}
		
		message Parameter {
			string parameterName = 1;
			string parameterValue = 2;
		}
	}
	
	message Age {
		uint64 old = 1;
		uint64 learningCycle = 2;
		repeated Environment environments = 3; 
	}

	message Environment {
		repeated Information informations = 1;
		string society = 2;
		
		message Information {
			string informationSource = 1;
			string connectedInputLayerName = 2;
		}
	}		
}